File: Data Structures & Algorithms/count-paths/submission-1.py
class Solution:
    def uniquePaths(self, m: int, n: int) -> int:
        if m == 1 and n == 1: return 1
        memo = {}
        def dp(i: int, j: int) -> int:
            if memo.get((i,j)):
                return memo[(i,j)]
            elif m-i == 1 and n-j == 0:
                memo[(i,j)] = 1
                return 1
            elif n-j == 1 and m-i == 0:
                memo[(i,j)] = 1
                return 1
            elif i>m or j>n:
                memo[(i,j)] = 0
                return 0
            else:
                memo[(i,j)] = dp(i+1,j) + dp(i,j+1)
                return memo[(i,j)]
        return dp(1,1)

File: Data Structures & Algorithms/house-robber/submission-1.cpp
class Solution {
public:
    int rob(vector<int>& nums) {
        // while you can technically skip over 3 houses,
        // it does not make sense.
        // you either skip one or two
        unordered_map<int,int> memo;
        return max(dfs(nums,0,memo), dfs(nums,1,memo));
    }
private:
    int dfs(vector<int>& nums, int ind, unordered_map<int,int>& memo){
        if (memo.count(ind)) return memo[ind];
        if (ind>=nums.size()) return 0;

        memo[ind] = nums[ind] + max(dfs(nums,ind+2,memo),dfs(nums,ind+3,memo));
        return memo[ind];       
    }
};
